ffi/firewood.go

```diff
@@ -347,11 +379,21 @@ func (db *Database) Close(ctx context.Context) error {
 		return fmt.Errorf("%w: %w", ctx.Err(), ErrActiveKeepAliveHandles)
 	}
 
+	db.stateLock.Lock()
+	defer db.stateLock.Unlock()
 	if err := getErrorFromVoidResult(C.fwd_close_db(db.handle)); err != nil {
 		return fmt.Errorf("unexpected error when closing database: %w", err)
 	}
 
+	db.handleLock.Lock()
 	db.handle = nil // Prevent double free
+	db.handleLock.Unlock()
 
 	return nil
 }
+
+func (db *Database) isClosed() bool {
+	db.handleLock.Lock()
+	defer db.handleLock.Unlock()
+	return db.handle == nil
```

Comment:
checking if the db is nil doesn't need to be locked if we're going to bail on if it is null; checking if the db is nil needs to be locked only if we are about to immediately use it. locking and then unlocking when the result is to return if nil is unnecessary. 

In other words, TOCTOU doesn't exist if there is no use. Otherwise, TOCTOU exists if you are using the handle after unlocking.